text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,339 @@
+    (a) Link cached object files into output binary, or
+    (b) Concatenate cached IR fragments, do optimization and codegen,
+        then link
+```
+
+
+## Miscellaneous Aspects
+
+### More fine-grained dependency tracking
+It would be possible to make dependency tracking aware of the kind of reference one item makes to another. If an item `A` mentions another item `B` only via some reference type (e.g. `&T`), then item `A` only needs to be updated if `B` is removed or `B` changes its 'sized-ness'. This is comparable to how forward declarations in C are handled. In the dependency graph this would mean that there are different kinds of edges that trigger for different kinds of changes to items.
+
+### Global Switches Influencing Codegen
+There are many compiler flags that change the way the generated code looks like, e.g. optimization and debuginfo levels. A simple strategy to deal with this would be to store the set of compiler flags used for building the cache and clearing the cache completely if another set of flags is used. Another option is to keep multiple caches, each for a different set of compiler flags (e.g. keeping both on disk, a 'debug build cache' and a 'release build cache').
+
+### Source Location Encoding
+The current `codemap` implementation in `rustc` stores everything in one global 'address space'. This is unsuited for incremental compilation since old versions of source files need to be removed from the codemap and new source files need to be added. This leads to a similar problem as the sequentially assigned AST node IDs.
+
+It would be better to store span information in byte-offsets relative to a given source file. Then modifying the codemap is possible without invalidating existing source locations stored somewhere in the cache.
+
+### Physical Cache Structure
+The cache could be kept somewhere within the output directory. Object code could be maintained in an `ar` archive so the linker can directly access it. Something similar might be possible for LLVM bitcode.
+
+### Automatic inter-function optimizations
+It should not be too hard to let the compiler keep track of which parts of the program change infrequently and then let it speculatively build object files with more than one function in them. For these aggregate object files inter-function LLVM optimizations could then be enabled, yielding faster object code at little additional cost. Other strategies for controlling cache granularity can be implemented in a similar fashion.
+
+### Parallelization
+If some care is taken in implementing the above concepts it should be rather easy to do translation and codegen in parallel for all items, since by design we already have (or can deterministically compute) all the information we need.
+
+# Drawbacks
+
+Implementing this will need a lot of work on the compiler architecture. But that's something that will be needed sooner or later anyway.
+
+# Alternatives
+
+I'd definitely like to hear about them.
```

Comment:
An alternative I have been thinking about as a long term solution is full incremental compilation (as opposed to incremental codegen) where we could  compile a single span, from parsing all the way to codegen. This is more useful for IDEs and similar tools, but would also give us more scope to incrementalise normal compilation. I envisaged generating much more thorough metadata for crates, enough that, for example, when a function is modified, that function could be compiled independently of the rest of the crate. This would require having all the type information we currently use for type checking in the metadata. (We would of course also need the object files and so forth used for incremental codegen).

The only way for this to be sane to implement would be if we had better representations of the compilers intermediate representations, and these could be serialised to make the metadata, rather than the current ad hoc approach (but this seems like a win from an architectural point of view too).